pkg/templateservicebroker/controller/templateprovision_controller.go

```diff
@@ -0,0 +1,142 @@
+}
+
+func NewTemplateProvisionController(kc *kclientset.Clientset, oc *client.Client, kubeconfig string) *TemplateProvisionController {
+	// QUERY: is it possible to avoid passing in kubeconfig here?  The factory
+	// object seems to be essential for bulk creating objects from a list.
+
+	loadingRules := &kclientcmd.ClientConfigLoadingRules{}
+	loadingRules.ExplicitPath = kubeconfig
+	loader := kclientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &kclientcmd.ConfigOverrides{})
+
+	c := TemplateProvisionController{factory: clientcmd.NewFactory(loader), kc: kc, oc: oc}
+
+	_, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return c.oc.TemplateProvisions(api.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return c.oc.TemplateProvisions(api.NamespaceAll).Watch(options)
+			},
+		},
+		&tsbapi.TemplateProvision{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.stateMachine(obj.(*tsbapi.TemplateProvision))
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				c.stateMachine(obj.(*tsbapi.TemplateProvision))
+			},
+			DeleteFunc: func(obj interface{}) {
+			},
+		},
+	)
+
+	return &c
+}
+
+func (c *TemplateProvisionController) Run(stop <-chan struct{}) {
+	c.controller.Run(stop)
+}
+
+func (c *TemplateProvisionController) stateMachine(tp *tsbapi.TemplateProvision) error {
+	var err error
+
+	switch tp.Status.State {
+	case tsbapi.TemplateProvisionStateNew:
+		tp.Status.State = tsbapi.TemplateProvisionStateCreating
+		_, err = c.oc.TemplateProvisions(tp.Namespace).Update(tp)
+
+	case tsbapi.TemplateProvisionStateCreating:
+		err = c.provision(tp)
+		if err == nil {
+			tp.Status.State = tsbapi.TemplateProvisionStateCreated
+			_, err = c.oc.TemplateProvisions(tp.Namespace).Update(tp)
+
+		} else {
+			glog.Errorf("templateprovisioncontroller: provision: %v", err)
+			tp.Status.State = tsbapi.TemplateProvisionStateFailed
+			c.oc.TemplateProvisions(tp.Namespace).Update(tp)
+		}
+	}
+
+	return err
+}
+
+func (c *TemplateProvisionController) provision(tp *tsbapi.TemplateProvision) error {
+	// TODO: impersonation
+
+	secret, err := c.kc.Secrets(tp.Namespace).Get(tp.Spec.SecretRef.Name)
+	if err != nil {
+		return err
+	}
+
+	template := &tp.Spec.Template
+
+	for i, param := range template.Parameters {
+		if value, ok := secret.Data[param.Name]; ok {
+			template.Parameters[i].Value = string(value)
+			template.Parameters[i].Generate = ""
+		}
+	}
+
+	template, err = c.oc.TemplateConfigs(tp.Namespace).Create(template)
+	if err != nil {
+		return err
+	}
+
+	errs := runtime.DecodeList(template.Objects, api.Codecs.UniversalDecoder())
+	if len(errs) > 0 {
+		return errs[0]
+	}
+
+	for _, obj := range template.Objects {
+		meta, _ := meta.Accessor(obj)
+		ref := meta.GetOwnerReferences()
+		ref = append(ref, metatypes.OwnerReference{
+			APIVersion: "v1",
```

Comment:
My takeaway from that conversation was that whatever goes in, it is going to have to change.  I could do `APIVersion: v1.SchemeGroupVersion.String()` where `v1` refers to `github.com/openshift/origin/pkg/template/api/v1`, but fundamentally that's just a complicated way of saying "v1".